sympy/calculus/util.py

```diff
@@ -709,6 +709,131 @@ def is_convex(f, *syms, **kwargs):
         return False
     return True
 
+
+def stationary_points(f, symbol, domain=S.Reals):
+    """
+    Returns the stationary points of a function (where derivative of the
+    function is 0) in the given domain.
+
+    Parameters
+    ==========
+
+    f : Expr
+        The concerned function.
+    symbol : Symbol
+        The variable for which the stationary points are to be determined.
+    domain : Interval
+        The domain over which the stationary points have to be checked.
+        If unspecified, S.Reals will be the default domain.
+
+    Examples
+    ========
+
+    >>> from sympy import Symbol, S, sin, log, pi, pprint, stationary_points
+    >>> from sympy.sets import Interval
+    >>> x = Symbol('x')
+
+    >>> stationary_points(1/x, x, S.Reals)
+    EmptySet()
+
+    >>> pprint(stationary_points(sin(x), x), use_unicode=False)
+              pi                              3*pi
+    {2*n*pi + -- | n in Integers} U {2*n*pi + ---- | n in Integers}
+              2                                2
+
+    >>> stationary_points(sin(x),x, Interval(0, 4*pi))
+    {pi/2, 3*pi/2, 5*pi/2, 7*pi/2}
+
+    """
+    from sympy import solveset, diff
+
+    if isinstance(domain, EmptySet):
+        return S.EmptySet
+
+    domain = continuous_domain(f, symbol, domain)
+    set = solveset(diff(f, symbol), symbol, domain)
```

Comment:
what about saddle points?

Reply:
What about multivariate functions?
What about non-differentiable functions?

Implementing this function is not as simple as it looks like if you want to handle all cases.

Reply:
This function is limited by `continuous_domain`. Since `continuous_domain` doesn't work good with multivariate and non-differentiable functions, `stationary_points` will have limited functionality too. `solveset` has this problem as well.

That is why I asked, if I should add failing tests as well. See this as well: https://github.com/sympy/sympy/pull/16473#pullrequestreview-220453891 and https://github.com/sympy/sympy/pull/16473#issuecomment-478353978

Reply:
While I agree that @Upabjojr has some valid points, I also think that this is a very good start for this. The functions can be extended to deal with the other cases as well eventually. (Similar to `solveset`, `integrate` etc.)